### PR TITLE
b/piboot.go: check EEPROM version for RPi4

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Check EEPROM version for RPi4 to make sure that tryboot is
supported. If the EEPROM need updating, return an error when
installing the kernel.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
When the EEPROM is old, this is what is shown in serial/display:

Captura desde 2022-04-04 15-54-12
Captura desde 2022-04-04 15-54-05

File: bootloader/piboot.go [hunk 1]
```diff
@@ -375,7 +383,55 @@ func (p *piboot) layoutKernelAssetsToDir(snapf snap.Container, dstDir string) er
...
+		return false
+	}
+
+	// The timestamp is seconds since the epoch, UTC time
+	eepromTs := binary.BigEndian.Uint32(buf)
+	// 2021-03-18 or more modern supports tryboot, see
+	// https://github.com/raspberrypi/rpi-eeprom/blob/master/firmware/release-notes.md#2021-04-19---promote-2021-03-18-from-latest-to-default---default
+	// The timestamp we compare with (jue 18 mar 2021 08:54:11 UTC) can be found with:
+	// $ strings pieeprom-2021-03-18.bin | grep BUILD_TIMESTAMP
+	if eepromTs < 1616057651 {
```
Comment (stolowski) [inline bootloader/piboot.go:401]: Should we do something like:
```
t := time.Date(2021, time.June, 18, 8, 54, 11, 0, time.UTC)
if eepromTs < t.Unix() {
```

instead so it's more self-documenting (and ensures correctness)? Having said that, I'm getting 1624006451 here.

Comment (stolowski) [reply bootloader/piboot.go:401]: I see, fair enough, I thought the timestamp from grep is actually in the human-readable format. In that case it makes sense to have it in seconds.

File: bootloader/piboot.go [hunk 2]
```diff
@@ -375,7 +383,55 @@ func (p *piboot) layoutKernelAssetsToDir(snapf snap.Container, dstDir string) er
...
 }
 
+func (p *piboot) eepromVersionSupportsTryboot() bool {
+	// To find out the EEPROM version we do the same as the
+	// rpi-eeprom-update script (see
+	// https://github.com/raspberrypi/rpi-eeprom/blob/master/rpi-eeprom-update)
+	buf, err := ioutil.ReadFile(rpi4EepromTimeStampPath)
+	if err != nil {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (mvo5) [reply bootloader/piboot.go:393]: Good point, I think we should at least log errors here but probably best to return an error

File: bootloader/piboot.go [hunk 3]
```diff
@@ -375,7 +383,55 @@ func (p *piboot) layoutKernelAssetsToDir(snapf snap.Container, dstDir string) er
...
+
+	return true
+}
+
 func (p *piboot) ExtractKernelAssets(s snap.PlaceInfo, snapf snap.Container) error {
+	if !p.prepareImageTime {
+		// If this is a RPi4, check first if EEPROM supports tryboot
+		if p.isRaspberryPi4() && !p.eepromVersionSupportsTryboot() {
+			return fmt.Errorf("your EEPROM does not support tryboot, please upgrade to a newer one before installing Ubuntu Core")
```
Comment (mvo5) [inline bootloader/piboot.go:431]: One nice touch would be if we could link to some forum topic with some more help. E.g. 
```
your EEPROM does not support tryboot, please upgrade to a newer one before installing Ubuntu Core, see http://forum.snapcraft.io/t/29455 for more details
```
(I created this unlisted page that probably needs some tweaks)

**Review by mvo5 — APPROVED** [review]
Thanks this is nice, some suggestios/nitpicks inline

File: bootloader/piboot.go [hunk 4]
```diff
@@ -375,7 +383,55 @@ func (p *piboot) layoutKernelAssetsToDir(snapf snap.Container, dstDir string) er
...
+	eepromTs := binary.BigEndian.Uint32(buf)
+	// 2021-03-18 or more modern supports tryboot, see
+	// https://github.com/raspberrypi/rpi-eeprom/blob/master/firmware/release-notes.md#2021-04-19---promote-2021-03-18-from-latest-to-default---default
+	// The timestamp we compare with (jue 18 mar 2021 08:54:11 UTC) can be found with:
+	// $ strings pieeprom-2021-03-18.bin | grep BUILD_TIMESTAMP
+	if eepromTs < 1616057651 {
+		return false
+	}
+
+	return true
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply bootloader/piboot.go]: Done

File: bootloader/piboot.go [hunk 5]
```diff
@@ -375,7 +383,55 @@ func (p *piboot) layoutKernelAssetsToDir(snapf snap.Container, dstDir string) er
...
+
+	// If old style codes (older than RPi2) or processor != BCM2711 (RPi4's SoC),
+	// this is not an RPi4. For details, see
+	// https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#raspberry-pi-revision-codes
+	boardInfo := binary.BigEndian.Uint32(buf)
+	if ((boardInfo>>23)&1) == 0 || ((boardInfo>>12)&0xF) != 3 {
+		return false
+	}
+
+	return true
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply bootloader/piboot.go]: Done

File: bootloader/piboot.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply bootloader/piboot.go]: Nice idea, I've changed the message accordingly.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the review, I've addressed the points raised now.

**Comment by mvo5** [comment]
Thanks! Hope squash merge is okay (for easier cherry picking) but it's just 4 commits, if you prefer them I can cherry-pick them individually

**Comment by alfonsosanchezbeato** [comment]
> Thanks! Hope squash merge is okay (for easier cherry picking) but it's just 4 commits, if you prefer them I can cherry-pick them individually

Sure, please squash merge if you prefer that.

**Pull request merged**